src/smtml/expr.ml

```diff
@@ -320,13 +303,66 @@ let exists vars body = binder Exists vars body
 
 let raw_unop ty op hte = make (Unop (ty, op, hte)) [@@inline]
 
+let is_val_real e =
+  match view e with
+  | Val (Real x) -> Float.is_finite x
+  | _ -> false
+
+let normalize_eq_or_ne op (ty', e1, e2) =
+  let is_real = is_val_real e1 && is_val_real e2 in
+  let make_relop lhs rhs = Relop (ty', op, lhs, rhs) in
+  let ty,ty2 = ty e1,ty e2 in
+  if Ty.equal ty ty2 then
+  match ty with
+  | Ty_bitv a ->
+     let binop = make (Binop (ty, Sub, e1, e2)) in
+     let zero = if a=32 then make (Val (Bitv (Bitvector.of_int32 0l)))
+                else make (Val (Bitv (Bitvector.of_int64 0L)))
+     in
+      make_relop binop zero
```

Comment:
```suggestion
  | Ty_bitv m ->
     let binop = make (Binop (ty, Sub, e1, e2)) in
     let zero = make (Val (Bitv (Bitvector.make Z.zero m))) in
      make_relop binop zero
```

No need for an if-statement, just say you want a zero with the size of the `Ty_bitv`